src/txn/txn_rollback_to_stable.c

```diff
@@ -0,0 +1,396 @@
+/*-
+ * Copyright (c) 2014-2017 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+#ifdef HAVE_TIMESTAMPS
+/*
+ * __txn_rollback_to_stable_lookaside_fixup --
+ *	Remove any updates that need to be rolled back from the lookaside file.
+ */
+static int
+__txn_rollback_to_stable_lookaside_fixup(
+    WT_SESSION_IMPL *session, wt_timestamp_t *rollback_timestamp)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	WT_ITEM las_addr, las_key, las_timestamp;
+	uint64_t las_counter, las_txnid, remove_cnt;
+	uint32_t las_id, session_flags;
+
+	conn = S2C(session);
+	cursor = NULL;
+	remove_cnt = 0;
+	session_flags = 0;		/* [-Werror=maybe-uninitialized] */
+	WT_CLEAR(las_timestamp);
+
+	__wt_las_cursor(session, &cursor, &session_flags);
+
+	/* Discard pages we read as soon as we're done with them. */
+	F_SET(session, WT_SESSION_NO_CACHE);
+
+	/* Walk the file. */
+	for (; (ret = cursor->next(cursor)) == 0; ) {
+		WT_ERR(cursor->get_key(cursor, &las_id, &las_addr, &las_counter,
+		    &las_txnid, &las_timestamp, &las_key));
+
+		/* Check the file ID so we can skip durable tables */
+		if (__bit_test(conn->stable_rollback_bitstring, las_id))
+			continue;
+
+		if (!__wt_timestamp_iszero(las_timestamp.data) &&
+		    __wt_timestamp_cmp(
+		    rollback_timestamp, las_timestamp.data) < 0) {
+			WT_ERR(cursor->remove(cursor));
+			++remove_cnt;
+		}
+	}
+	WT_ERR_NOTFOUND_OK(ret);
+err:	WT_TRET(__wt_las_cursor_close(session, &cursor, session_flags));
+	/*
+	 * If there were races to remove records, we can over-count. Underflow
+	 * isn't fatal, but check anyway so we don't skew low over time.
+	 */
+	if (remove_cnt > conn->las_record_cnt)
+		conn->las_record_cnt = 0;
+	else if (remove_cnt > 0)
+		(void)__wt_atomic_sub64(&conn->las_record_cnt, remove_cnt);
+
+	F_CLR(session, WT_SESSION_NO_CACHE);
+
+	return (ret);
+}
+
+/*
+ * __txn_abort_newer_update --
+ *	Abort updates in an update change with timestamps newer than the
+ *	rollback timestamp.
+ */
+static void
+__txn_abort_newer_update(WT_SESSION_IMPL *session,
+    WT_UPDATE *upd, wt_timestamp_t *rollback_timestamp)
+{
+	WT_UPDATE *next_upd;
+	bool aborted_one;
+
+	aborted_one = false;
+	for (next_upd = upd; next_upd != NULL; next_upd = next_upd->next) {
+		/* Only updates with timestamps will be aborted */
+		if (!__wt_timestamp_iszero(&next_upd->timestamp) &&
+		    __wt_timestamp_cmp(
+		    rollback_timestamp, &next_upd->timestamp) < 0) {
```

Comment:
Is the `__wt_timestamp_iszero` check needed?  (i.e., can a zero timestamp ever be greater than the `rollback_timestamp`?